src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/MessagePackHubProtocol.java

```diff
@@ -0,0 +1,580 @@
+                        break;
+                    case CLOSE:
+                        hubMessages.add(createCloseMessage(unpacker, itemCount));
+                        break;
+                    default:
+                        break;
+                }
+            }
+        } catch (MessagePackException | IOException ex) {
+            throw new RuntimeException("Error reading MessagePack data.", ex);
+        }
+
+        return hubMessages.toArray(new HubMessage[hubMessages.size()]);
+    }
+    
+    @Override
+    public String writeMessage(HubMessage hubMessage) {
+        HubMessageType messageType = hubMessage.getMessageType();
+        
+        try {
+            switch (messageType) {
+                case INVOCATION:
+                    return writeInvocationMessage((InvocationMessage) hubMessage);
+                case STREAM_ITEM:
+                    return writeStreamItemMessage((StreamItem) hubMessage);
+                case COMPLETION:
+                    return writeCompletionMessage((CompletionMessage) hubMessage);
+                case STREAM_INVOCATION:
+                    return writeStreamInvocationMessage((StreamInvocationMessage) hubMessage);
+                case CANCEL_INVOCATION:
+                    return writeCancelInvocationMessage((CancelInvocationMessage) hubMessage);
+                case PING:
+                    return writePingMessage((PingMessage) hubMessage);
+                case CLOSE:
+                    return writeCloseMessage((CloseMessage) hubMessage);
+                default:
+                    throw new RuntimeException(String.format("Unexpected message type: %d", messageType.value));
+            }
+        } catch (MessagePackException | IOException ex) {
+            throw new RuntimeException("Error writing MessagePack data.", ex);
+        }
+    }
+        
+    private HubMessage createInvocationMessage(MessageUnpacker unpacker, InvocationBinder binder, int itemCount) throws IOException {
+        Map<String, String> headers = readHeaders(unpacker);
+        String invocationId = unpacker.unpackString();
+        
+        // For MsgPack, we represent an empty invocation ID as an empty string,
+        // so we need to normalize that to "null", which is what indicates a non-blocking invocation.
+        if (invocationId == null || invocationId.isEmpty()) {
+            invocationId = null;
+        }
+            
+        String target = unpacker.unpackString();
+        
+        Object[] arguments = null;
+        try {
+            List<Class<?>> types = binder.getParameterTypes(target);
+            arguments = bindArguments(unpacker, types);
+        } catch (Exception ex) {
+            return new InvocationBindingFailureMessage(invocationId, target, ex);
+        }
+        
+        // NOTE - C# client accounts for possibility of streamId array being absent, while spec claims it's required - which?
```

Comment:
The C# client checks to see if the StreamID array exists, though the spec claims it's required: https://github.com/dotnet/aspnetcore/blob/0d156487aa27df65d68fc7fa31e7b6cb9f4d7581/src/SignalR/common/Protocols.MessagePack/src/Protocol/MessagePackHubProtocolWorker.cs#L90-L95. Do I need to do a similar check?

Reply:
It's probably a good idea to add a check. It's required for the latest version of the protocol, but I'm pretty sure we check for backwards compatibility with clients using older versions of the protocol. A test for this would also be a good idea.

Reply:
> the spec claims it's required

Maybe need to update the language, where was it saying it's required?

In order to reduce bandwidth we can avoid sending the stream ID array when there aren't any streams, also as Stephen says, this preserves backwards compatibility with older clients.

Reply:
Well, maybe I'm wrong - the MessagePack portion of the spec shows 0-length StreamID arrays when it's absent: https://www.javadoc.io/doc/com.google.code.gson/gson/latest/com.google.gson/com/google/gson/Gson.html

But the JsonProtocol portion claims it's optional: https://github.com/dotnet/aspnetcore/blob/master/src/SignalR/docs/specs/HubProtocol.md#invocation-message-encoding

I'll update to account for it being optional